paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/AbstractSparkInternalRow.java

```diff
@@ -251,11 +251,21 @@ public boolean equals(Object o) {
             return false;
         }
         AbstractSparkInternalRow that = (AbstractSparkInternalRow) o;
-        return Objects.equals(rowType, that.rowType) && Objects.equals(row, that.row);
+        try {
+            return Objects.equals(rowType, that.rowType) && Objects.equals(row, that.row);
+        } catch (Exception e) {
+            // The underlying row may not support equals or hashcode, e.g., `ProjectedRow`,
+            // to be safe, we fallback to super.
+            return super.equals(o);
```

Comment:
The behavior after fallback is not correct, it only compares instances.

Reply:
This is mainly used to avoid exception. If we want to make behavior correct, how about doing a deep copy and then compare ?

Reply:
Maybe we should implement a method to equals with row type (invoke row.getXXX).

Reply:
Maybe introduce `InternalRowUtils.equals(InternalRow row1, InternalRow row2, RowType rowType)`

Reply:
addressed